design-proposals/volume-claim-templates.md

```diff
@@ -0,0 +1,117 @@
+A new API will be created in the CDI API library which will receive a partial PVC template and using existing CDI logic will return a complete PVC spec.  
+The provided PVC template in the VolumeClaimTemplate should have a source that can be handled by the new existing CDI volume populators API (DataSourceRef with CDI CR source) or a standalone k8s way (user costume populator, empty disk etc..)
+
+A virtctl command called `create-pvc` can be added. Given minimal flags such as name, size and source this command can generate and create a PVC. The command will use the CDI API similarly as will be done in the VM controller.
+
+## API Examples
+### VM with VolumeClaimTemplates, VolumeImportSource as volume source
+
+This example will create a VolumeImportSource along side with a VM. Once the VM is created we will generate a PVC spec with the optimal VolumeMode and AccessModes. This PVC will be created and populated with the provided image in the VolumeImportSource url.
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: VolumeImportSource
+metadata:
+  name: my-import-source
+spec:
+  source:
+      http:
+         url: "https://download.cirros-cloud.net/0.4.0/cirros-0.4.0-x86_64-disk.img"
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: cirros
+spec:
+  running: true
+  volumeClaimTemplates:
+  - metadata:
+      name: cirros-pvc
+    spec:
+    dataSourceRef:
+      apiGroup: cdi.kubevirt.io
+      kind: VolumeImportSource
+      name: my-import-source
+    resources:
+      requests:
+        storage: 1Gi
+  template:
+    spec:
+      domain:
+        devices: {}
+      volumes:
+      - persistentVolumeClaim:
+          name: cirros-pvc
+        name: cirros-disk
+```
+
+#### Open question:
+With datavolumestemplate one of the downsides was that we couldn't use generateName of the datavolume name.
+It will be cool if we could use it for the PVC name. Issue: we need to mention the PVC name in the vm.spec.template.volumes part when writing the VM spec. Any idea how to go past that? update the PVC name there after the PVC is created sounds like not the best solution. WDYT?
+
+### Virtctl create-pvc command
+
+Example for virtctl command uses:
+```bash
+$ virtctl create-pvc --generate-name=my-pvc- --size=1Gi
+```
+This command will create a PVC with a generated name with "my-pvc-" prefix. The PVC size will be 1Gi, it will have empty disk image and AccessModes and VolumeMode optimal for the defined default storage class.
+
+```bash
+$ virtctl create-pvc --name=my-pvc- --volume-import-source=my-import-source --size=1Gi --storage-class=rook-ceph-block
+```
+This command will create a PVC named "my-pvc" with AccessModes and VolumeMode optimal for rook-ceph-block. The PVC will be populated by the data source provided in VolumeImportSource CR called "my-import-source".
+
+```bash
+$ virtctl create-pvc --name=my-pvc --volume-clone-source=fedora
```

Comment:
It's not clear how rook-ceph-block comes into play specifically here.  If the default SC is ceph-rbd then sure.  Maybe you want a --storage-class option to the command in order to better illustrate the point.  Instead of --volume-clone-source how about we make a single option --datasourceref that allows you to specify any populator CR?

Reply:
regarding the storage-class I think you got confused with the order of command and under it the description, the rook-ceph-block description refers to the command where the user provides --storage-class flag specifically.
regarding the datasourceref, you mean to provide both:
--datasourceref-kind=volumeCloneSource
--datasourceref-name=my-clone-source
?
It does give you the option to use costume populator, but requires 2 flags instead of 1.